packages/backend/src/index.ts

```diff
@@ -124,13 +231,44 @@ async function main() {
   apiRouter.use('/auth', await auth(authEnv));
   apiRouter.use('/azure-devops', await azureDevOps(azureDevOpsEnv));
   apiRouter.use('/search', await search(searchEnv));
-  apiRouter.use('/techdocs', await techdocs(techdocsEnv));
-  apiRouter.use('/todo', await todo(todoEnv));
+
+  /**
+   * Injecting a child container to the module. Child module should populate missing implementations
+   * to the container or default to parent container implementations if such exists
+   */
+  apiRouter.use(
+    '/badges',
+    await badges({ ...badgesEnv, container: inversifyContainer.createChild() }),
```

Comment:
Very much a nit at this point, but I actually think we would likely end up with completely separate containers for each plugin since many instances in the plugin environment are per-plugin. We don't really want plugins to interact with each other through code anyway, so I think that isolation could be a benefit. That's ofc only if it's still as easy to set it up, we wouldn't want to add a bunch of boiler plate to support it.

Reply:
Would be useful to see how factories might support some form of the `forPlugin` pattern that some of these utilities currently use. That is, the derived types needing a plugin ID as its input to be able to create itself. Maybe that plugin ID (or rather, a `BackendPlugin` that has an `id`?) is just another value provided in the container?

Reply:
With separate containers I think `forPlugin` can happen before instantiation of the container, kinda this:

```ts
const fooManager = FooManager.create();

function createPluginContainer(pluginId: string) {
  return new Container().withFactory('FooApi', () => fooManager.forPlugin(pluginId))
}
```

Something we realized with the ApiRef stuff is that branching of instances within the DI framework adds **a lot** of complexity.

Although it could also be more like this, which is perhaps what you meant to begin with:

```ts
// somewhere else
class FooManager implements FooApi {
  factory(@inject('PluginId') pluginId) {
    return new FooManager(pluginId)
  }
}

function createPluginContainer(pluginId: string) {
  return new Container()
    .withFactory('PluginId', () => pluginId)
    .withFactory('FooApi', FooManager.factory)
}
```

That ofc only works when you don't have the need to share a root instance between the plugins

Reply:
The latter, yes. In the first one you're responsible for making the entire container explicitly, in the latter it's easier to point-replace single pieces of the chain. If I'm reading the suggestion right

Reply:
I'm essentially saying that `FooManager` would be part of the parent container from the framework side. And then there's a default implementation of the factory for `PluginFooManager` that requires a plugin ID - which was also supplied by the backend core into the container before entering the plugin's actual init code.

But maybe this is actually a use case that isn't super common to need 🤔

Reply:
So `FooManager` would be a root thing that should never be accessed by plugins, and `PluginFooManager` is the actual thing they wanna use and we'll add it in manually to each child container so that the instances aren't shared? or whatever other way we can make sure that it's possible to have multiple different `PluginFooManager`s at once

Reply:
We could possibly use Inversify's scopes to our advantage here. I have a feeling that we do need to functionality to be able to traverse up to a parent container to get instances in cases where an implementation is provided by the backend instead of within the module itself.